v4-migrator: VULNERABILITY table fails to be converted

### Current Behavior

While testing migration and after overcoming previously reported issues (thanks for the prompt fixes!) I encountered another error during migration of the `VULNERABILITY` table:
```
2026-06-01 12:46:59.074 INFO  o.d.v4migrator.load.LoadPhase - Loading VULNERABILITY into v5
org.jdbi.v3.core.statement.UnableToExecuteStatementException: org.postgresql.util.PSQLException: ERROR: column "SEVERITY" does not exist
  Detail: There is a column named "SEVERITY" in table "VULNERABILITY", but it cannot be referenced from this part of the query.
  Position: 1251 [statement:"INSERT INTO "VULNERABILITY" (
    "ID"
  , "CREATED"
  , "CREDITS"
  , "CVSSV2BASESCORE"
  , "CVSSV2EXPLOITSCORE"
  , "CVSSV2IMPACTSCORE"
  , "CVSSV2VECTOR"
  , "CVSSV3BASESCORE"
  , "CVSSV3EXPLOITSCORE"
  , "CVSSV3IMPACTSCORE"
  , "CVSSV3VECTOR"
  , "CVSSV4SCORE"
  , "CVSSV4VECTOR"
  , "CWES"
  , "DESCRIPTION"
  , "DETAIL"
  , "FRIENDLYVULNID"
  , "OWASPRRBUSINESSIMPACTSCORE"
  , "OWASPRRLIKELIHOODSCORE"
  , "OWASPRRTECHNICALIMPACTSCORE"
  , "OWASPRRVECTOR"
  , "PATCHEDVERSIONS"
  , "PUBLISHED"
  , "RECOMMENDATION"
  , "REFERENCES"
  , "SEVERITY"
  , "SOURCE"
  , "SUBTITLE"
  , "TITLE"
  , "UPDATED"
  , "UUID"
  , "VULNID"
  , "VULNERABLEVERSIONS"
)
SELECT "ID"
     , "CREATED"
     , "CREDITS"
     , "CVSSV2BASESCORE"
     , "CVSSV2EXPLOITSCORE"
     , "CVSSV2IMPACTSCORE"
     , "CVSSV2VECTOR"
     , "CVSSV3BASESCORE"
     , "CVSSV3EXPLOITSCORE"
     , "CVSSV3IMPACTSCORE"
     , "CVSSV3VECTOR"
     , "CVSSV4SCORE"
     , "CVSSV4VECTOR"
     , "CWES"
     , "DESCRIPTION"
     , "DETAIL"
     , "FRIENDLYVULNI[...]", arguments:{}]
        at org.jdbi.v3.core.statement.SqlStatements.handleException(SqlStatements.java:493)
        at org.jdbi.v3.core.statement.SqlStatement.internalExecute(SqlStatement.java:1824)
        at org.jdbi.v3.core.result.ResultProducers.lambda$returningUpdateCount$0(ResultProducers.java:49)
        at org.jdbi.v3.core.statement.Update.execute(Update.java:79)
        at org.jdbi.v3.core.statement.Update.execute(Update.java:58)
        at org.jdbi.v3.core.Handle.execute(Handle.java:354)
        at org.jdbi.v3.core.Handle.execute(Handle.java:367)
        at org.dependencytrack.v4migrator.load.LoadPhase.lambda$8(LoadPhase.java:276)
        at org.jdbi.v3.core.transaction.LocalTransactionHandler$BoundLocalTransactionHandler.inTransaction(LocalTransactionHandler.java:229)
        at org.jdbi.v3.core.Handle.inTransaction(Handle.java:719)
        at org.jdbi.v3.core.Jdbi.lambda$inTransaction$0(Jdbi.java:457)
        at org.jdbi.v3.core.Jdbi.withHandle(Jdbi.java:423)
        at org.jdbi.v3.core.Jdbi.inTransaction(Jdbi.java:457)
        at org.dependencytrack.v4migrator.load.LoadPhase.loadOne(LoadPhase.java:276)
        at org.dependencytrack.v4migrator.load.LoadPhase.run(LoadPhase.java:77)
        at org.dependencytrack.v4migrator.cli.LoadCommand.execute(LoadCommand.java:37)
        at org.dependencytrack.v4migrator.cli.AbstractMigratorCommand.call(AbstractMigratorCommand.java:62)
        at org.dependencytrack.v4migrator.cli.AbstractMigratorCommand.call(AbstractMigratorCommand.java:1)
        at picocli.CommandLine.executeUserObject(CommandLine.java:2031)
        at picocli.CommandLine.access$1500(CommandLine.java:148)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2469)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2461)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2423)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2425)
        at picocli.CommandLine.execute(CommandLine.java:2174)
        at org.dependencytrack.v4migrator.V4Migrator.main(V4Migrator.java:49)
Caused by: org.postgresql.util.PSQLException: ERROR: column "SEVERITY" does not exist
  Detail: There is a column named "SEVERITY" in table "VULNERABILITY", but it cannot be referenced from this part of the query.
  Position: 1251
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2904)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2589)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:431)
        at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:534)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:437)
        at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:196)
        at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:182)
        at org.jdbi.v3.core.statement.SqlLoggerUtil.wrap(SqlLoggerUtil.java:32)
        at org.jdbi.v3.core.statement.SqlStatement.internalExecute(SqlStatement.java:1822)
        ... 25 more
```

### Steps to Reproduce

1. Start a migration
2. Observe the error when reaching the `VULNERABILITY` table

### Expected Behavior

Migration of `VULNERABILITY` table does not cause errors.

**Note**:
This is still the same run that resulted in reporting #6217 and #6218, fixing things manually along the way. I think it is very unlikely to be a side-effect of those (seemingly being an SQL error) but I do not want to rule it out entirely.

### Dependency-Track Version

5.x

### Browser

Google Chrome

### Checklist

- [x] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/main/CONTRIBUTING.md#filing-issues)
- [x] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this defect was already reported

--- a/support/v4-migrator/src/main/java/org/dependencytrack/v4migrator/TableRegistry.java
+++ b/support/v4-migrator/src/main/java/org/dependencytrack/v4migrator/TableRegistry.java
@@ -1530,7 +1530,10 @@ public final class TableRegistry {
           , "LOG_SUCCESSFUL_PUBLISH"      boolean
           , "MESSAGE"                     varchar(1024)
           , "NAME"                        varchar(255) NOT NULL
-          , "NOTIFICATION_LEVEL"          notification_level
+          -- Stored as text + CHECK rather than the native "notification_level" enum so a
+          -- target-side DROP SCHEMA public CASCADE does not transitively drop this column.
+          -- Keep the value list in sync with migration/.../V202605022031__init.sql.
+          , "NOTIFICATION_LEVEL"          varchar(255) CHECK ("NOTIFICATION_LEVEL" IN ('INFORMATIONAL', 'WARNING', 'ERROR'))
           , "NOTIFY_CHILDREN"             boolean
           , "NOTIFY_ON"                   text[]
           , "PUBLISHER"                   bigint
@@ -1550,7 +1553,7 @@ public final class TableRegistry {
              , r."LOG_SUCCESSFUL_PUBLISH"
              , r."MESSAGE"
              , r."NAME"
-             , CASE WHEN r."NOTIFICATION_LEVEL" IS NULL THEN NULL ELSE r."NOTIFICATION_LEVEL"::notification_level END
+             , r."NOTIFICATION_LEVEL"
              , r."NOTIFY_CHILDREN"
              , CASE WHEN r."NOTIFY_ON" IS NULL THEN NULL ELSE string_to_array(r."NOTIFY_ON", ',') END
              , m.canonical_id
@@ -1595,7 +1598,7 @@ public final class TableRegistry {
              , "LOG_SUCCESSFUL_PUBLISH"
              , "MESSAGE"
              , "NAME"
-             , "NOTIFICATION_LEVEL"
+             , CASE WHEN "NOTIFICATION_LEVEL" IS NULL THEN NULL ELSE "NOTIFICATION_LEVEL"::notification_level END
              , "NOTIFY_CHILDREN"
              , "NOTIFY_ON"
              , "PUBLISHER"
@@ -3235,7 +3238,11 @@ public final class TableRegistry {
           , "PUBLISHED"                   timestamp with time zone
           , "RECOMMENDATION"              text
           , "REFERENCES"                  text
-          , "SEVERITY"                    severity
+          -- Stored as text + CHECK rather than the native "severity" enum so a target-side
+          -- DROP SCHEMA public CASCADE (a common manual reset between load attempts) does
+          -- not transitively drop this column. Keep the value list in sync with
+          -- migration/.../V202605022031__init.sql.
+          , "SEVERITY"                    varchar(255) CHECK ("SEVERITY" IN ('UNASSIGNED', 'INFO', 'LOW', 'MEDIUM', 'HIGH', 'CRITICAL'))
           , "SOURCE"                      varchar(255) NOT NULL
           , "SUBTITLE"                    varchar(255)
           , "TITLE"                       varchar(255)
@@ -3304,7 +3311,9 @@ public final class TableRegistry {
              , "PUBLISHED"
              , "RECOMMENDATION"
              , "REFERENCES"
-             , "SEVERITY"::severity
+             -- v4 allows NULL "SEVERITY"; treat that as the canonical UNASSIGNED sentinel
+             -- so the v5 column is never NULL.
+             , COALESCE("SEVERITY", 'UNASSIGNED')
              , "SOURCE"
              , "SUBTITLE"
              , "TITLE"
@@ -3376,7 +3385,7 @@ public final class TableRegistry {
              , "PUBLISHED"
              , "RECOMMENDATION"
              , "REFERENCES"
-             , "SEVERITY"
+             , "SEVERITY"::severity
              , "SOURCE"
              , "SUBTITLE"
              , "TITLE"

--- a/support/v4-migrator/src/test/java/org/dependencytrack/v4migrator/VulnerabilityMigrationIT.java
+++ b/support/v4-migrator/src/test/java/org/dependencytrack/v4migrator/VulnerabilityMigrationIT.java
@@ -89,6 +89,13 @@ class VulnerabilityMigrationIT {
                 INSERT INTO "VULNERABILITY" ("ID", "UUID", "VULNID", "SOURCE", "SEVERITY")
                 VALUES (2, 'not-a-uuid', 'CVE-2024-0002', 'NVD', 'LOW')
                 """);
+
+            // NULL SEVERITY in v4 (legacy pre-ComputeSeveritiesChange data): must land
+            // as UNASSIGNED in v5.
+            h.execute("""
+                INSERT INTO "VULNERABILITY" ("ID", "UUID", "VULNID", "SOURCE", "SEVERITY")
+                VALUES (3, '00000000-0000-0000-0000-000000000003', 'CVE-2024-0003', 'NVD', NULL)
+                """);
         });
 
         runPipeline();
@@ -114,9 +121,11 @@ class VulnerabilityMigrationIT {
                       FROM dt_v4_migration.vulnerability_canonical_id_map
                      ORDER BY orig_id
                     """).mapToMap().list());
-        assertThat(map).extracting("orig_id", "canonical_id").containsExactly(tuple(1L, 1L));
+        assertThat(map).extracting("orig_id", "canonical_id")
+            .containsExactly(tuple(1L, 1L), tuple(3L, 3L));
 
-        // v5 row: SEVERITY is the native enum, UUID is native, EPSS columns are gone.
+        // v5 rows: SEVERITY is the native enum, UUID is native, EPSS columns are gone,
+        // and the NULL-SEVERITY row has been normalized to UNASSIGNED.
         final List<Map<String, Object>> rows = target.jdbi().withHandle(h ->
             h.createQuery("""
                     SELECT "ID", "UUID", "VULNID", "SOURCE",
@@ -126,7 +135,7 @@ class VulnerabilityMigrationIT {
                       FROM "VULNERABILITY"
                      ORDER BY "ID"
                     """).mapToMap().list());
-        assertThat(rows).hasSize(1);
+        assertThat(rows).hasSize(2);
         assertThat(rows.get(0))
             .containsEntry("id", 1L)
             .containsEntry("uuid", UUID.fromString("00000000-0000-0000-0000-000000000001"))
@@ -138,6 +147,10 @@ class VulnerabilityMigrationIT {
             .containsEntry("references", "https://example/ref")
             .containsEntry("cwes", "79,89");
         assertThat((BigDecimal) rows.get(0).get("cvssv3basescore")).isEqualByComparingTo("7.5");
+        assertThat(rows.get(1))
+            .containsEntry("id", 3L)
+            .containsEntry("vulnid", "CVE-2024-0003")
+            .containsEntry("severity_text", "UNASSIGNED");
 
         // EPSS columns do not exist on v5 VULNERABILITY.
         final List<Map<String, Object>> epssCols = target.jdbi().withHandle(h ->

